plugin/completion.py

```diff
@@ -205,32 +199,94 @@ def on_query_completions(self, prefix: str, locations: List[int]) -> Optional[su
             return None
         self.manager.documents.purge_changes(self.view)
         completion_list = sublime.CompletionList()
+        self._request_in_flight = True
         client.send_request(
             Request.complete(text_document_position_params(self.view, locations[0])),
-            lambda res: self.handle_response(res, completion_list, self.view.change_id()),
+            lambda res: self.handle_response(res, completion_list),
             lambda res: self.handle_error(res, completion_list))
         return completion_list
 
-    def handle_response(self, response: Optional[Union[dict, List]],
-                        completion_list: sublime.CompletionList, change_id: Any) -> None:
+    def format_completion(self, item: dict) -> sublime.CompletionItem:
+        item_kind = item.get("kind")
+        if item_kind:
+            kind = completion_kinds.get(item_kind, sublime.KIND_AMBIGUOUS)
+        else:
+            kind = sublime.KIND_AMBIGUOUS
+
+        if item.get("deprecated", False):
+            kind = (kind[0], '⚠', "⚠ {} - Deprecated".format(kind[2]))
+
+        text_edit = item.get("textEdit")
+        if text_edit:
+            r = text_edit["range"]
+            start, end = r["start"], r["end"]
+            # From the LSP spec: the range of the edit must be a single line range and it must contain the position at
+            # which completion has been requested.
+            row, start_col_utf16, end_col_utf16 = start["line"], start["character"], end["character"]
+            # The TextEdit from the language server might apply to an old version of the buffer. The user may have
+            # entered more characters in the meantime.
+            global _changes_since_sent
+            for change in _changes_since_sent:
+                # Thus, adjust the range for each buffer change.
+                start_col_utf16, end_col_utf16 = transform_region(row, start_col_utf16, end_col_utf16, change)
+            # We don't have to use region_to_range. That function clamps the unreliable input, but having unreliable
```

Comment:
Can we somehow apply _changes_since_sent to only the selected completion item (in the LspSelectCompletionItemCommand), instead of all completion items. 

I don't see a point in applying the transform_region to all items, because only one will be chosen.

Reply:
To sum up the problem that this PR solves:

If the user types:

```
1| h|
```
-> completions `request_1` is sent.

User continues typing while, but the request_1 is not finished

```
1| he|
```

<- responses from request_1 arrived, users sees completion items in AC.

-> request_2 has been send (lets say it will take 5sec for the second request to complete, it is not a real case but lets imagine that this happens)

The user sees completion items from request_1
that were calculated when the view state was:
```
1| h|
```

but the current view state is:
```
1| he|
```

When the select a completion item,
because the view state has changed. 
The text edit wont be applied correctly.

<hr>

There are two solutions to this problem
one of them being your solution and that is:

When a request is send.
You record each edit that happened (with on_text_changed).
This way you pick that the letter `e` was typed.
And call transform_region on **all completion items** returned from the server
to change the TextEdit range to include all text changes. 
Keep in mind that there could be 1000 of completion items, but the user will only select one. 
So you did 999 of unnecessary calls to transform_region. (which I personally don't find that bad, but I want to point out that)

But another solution is to just bring views state to the state it was when the request was made.

So we need the current view state: 
```
1| he|
```

to bring to the state it was when the request was made ant that is:
```
1| h|
```

That way we can avoid transforming TextEdit-

This was all I had to say.
Functionality wise this PR does the job
and it fixes #957

Reply:
If I move the adjustment of the text edit range to the LspInsertCompletionItem command, then my test fails:

```
======================================================================
FAIL: test_nontrivial_text_edit_removal (test_completion.QueryCompletionsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/raoul/.config/sublime-text-3/Packages/UnitTesting/unittesting/core/py33/case.py", line 28, in _executeTestPart
    yield from deferred
  File "/home/raoul/.config/sublime-text-3/Packages/LSP/tests/test_completion.py", line 509, in test_nontrivial_text_edit_removal
    self.assertEqual(self.read_file(), '#include <uchar.h>')
AssertionError: '#include <uchar.h>>' != '#include <uchar.h>'
- #include <uchar.h>>
?                   -
+ #include <uchar.h>
```

I am guessing this is because `transform_region` won't work if the prefix string is already removed (which it is if we're inside LspSelectCompletionCommand).

> Keep in mind that there could be 1000 of completion items, but the user will only select one. So you did 999 of unnecessary calls to transform_region.

Yes :/ no way around it.

Well, with the new `details` field at the bottom we're going to parse 1000 markdown strings into HTML anyway ;)

Reply:
Besides a quality lang server will stop at ~200 items and set `isIncomplete` to `True`.

Reply:
I invite you to try moving the logic to LspSelectCompletionCommand, and convince yourself that it will be impossible to do so.

Reply:
>  1000 markdown strings into HTML anyway

True :D